bulldozer/signals.go

```diff
@@ -21,62 +21,186 @@ import (
+	return len(*signal) > 0
+}
+
+func (signal *BranchPatterns) Enabled() bool {
+	return len(*signal) > 0
+}
+
+func (signal *MaxCommits) Enabled() bool {
+	return *signal > 0
 }
 
 func (s *Signals) Enabled() bool {
-	size := 0
-	size += len(s.Labels)
-	size += len(s.CommentSubstrings)
-	size += len(s.Comments)
-	size += len(s.PRBodySubstrings)
-	size += len(s.Branches)
-	size += len(s.BranchPatterns)
-	return size > 0
+	return s.Labels.Enabled() ||
+		s.CommentSubstrings.Enabled() ||
+		s.Comments.Enabled() ||
+		s.PRBodySubstrings.Enabled() ||
+		s.Branches.Enabled() ||
+		s.BranchPatterns.Enabled() ||
+		s.MaxCommits.Enabled()
 }
 
-// Matches returns true if the pull request meets one or more signals. It also
+// MatchesAll returns true if the pull request matches ALL of the signals. It also
+// returns a description of the match status. The tag argument appears
+// in this description and indicates the behavior (trigger, ignore) this
+// set of signals is associated with.
+func (s *Signals) MatchesAll(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+	signals := []Signal{
+		&s.Labels,
+		&s.CommentSubstrings,
+		&s.Comments,
+		&s.PRBodySubstrings,
+		&s.Branches,
+		&s.BranchPatterns,
+		&s.MaxCommits,
+	}
+
+	for _, signal := range signals {
+		matches, _, err := signal.Matches(ctx, pullCtx, tag)
+		if err != nil {
+			return false, "", err
+		}
+
+		if signal.Enabled() && !matches {
+			return false, fmt.Sprintf("pull request does not match all %s signals", tag), nil
+		}
+	}
+
+	return true, fmt.Sprintf("pull request matches all %s signals", tag), nil
+}
+
+// MatchesAny returns true if the pull request meets one or more signals. It also
 // returns a description of the signal that was met. The tag argument appears
 // in this description and indicates the behavior (trigger, ignore) this
 // set of signals is associated with.
-func (s *Signals) Matches(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+func (s *Signals) MatchesAny(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+	signals := []Signal{
+		&s.Labels,
+		&s.CommentSubstrings,
+		&s.Comments,
+		&s.PRBodySubstrings,
+		&s.Branches,
+		&s.BranchPatterns,
+	}
+
+	for _, signal := range signals {
+		matches, description, err := signal.Matches(ctx, pullCtx, tag)
+		if err != nil {
+			return false, "", err
+		}
+
+		if matches {
+			return true, description, nil
+		}
+	}
+
+	return false, fmt.Sprintf("pull request does not match the %s", tag), nil
+}
+
+// Matches Determines which label signals match the given PR. It returns:
+// - A boolean to indicate if a signal matched
+// - A description of the first matched signal
+func (signal *Labels) Matches(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
 	logger := zerolog.Ctx(ctx)
 
+	if !signal.Enabled() {
+		logger.Debug().Msgf("No label signals have been provided to match against")
```

Comment:
Even if you keep the structure of `MatchesAll` the same, I think we may want to remove these log messages to reduce noise. I'm not sure how useful these will be to someone debugging a problem, especially when only a few signal types are in use at any given time.